Import feature: show the differences of an element when it is updated by an import (and refactor)

MyPyDavid opened the pull request:
<!-- These comments are hidden when you submit the pull request, so you do not need to remove them. -->

<!--- Provide a general summary of your changes in the Title above --->

## Description
<!--- Describe your changes in detail --->

<!--- This project only accepts pull requests related to open issues --->
<!--- If suggesting a new feature or change, please discuss it in an issue first --->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce --->
<!--- Please link to the issue here: --->

This PR implements a text difference viewer for the import of RDMO content elements (catalogs, optionsets, etc..). 
It uses the [react-diff-viewer-continued](https://www.npmjs.com/package/react-diff-viewer-continued) to display the changes during an import. In the back-end a sort of "track changes" function tracks the current and new state of each element which is returned in the json response.  
The code architecture of the import is refactored into one central import function under `management` with helpers for each type of element. 


Related issues: #468, #469

Includes bug fix for: #998 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? --->

## How has this been tested?
<!--- Please describe in detail how you tested your changes. --->
<!--- Include details of your testing environment, and the tests you ran to --->
<!--- see how your change affects other areas of the code, etc. --->

## Screenshots (if appropriate)
Currently: 

![management-import-optionsets-1-changes](https://github.com/rdmorganiser/rdmo/assets/13996213/3b07797c-f1b6-4466-98dd-bd5153d01a36)

<details>
<summary>previous version</summary>
![management-import-optionsets-1-changes](https://github.com/rdmorganiser/rdmo/assets/13996213/e5fcb01c-bc8e-4c86-9b89-4318fd015c47)
</details>

<details>
<summary>older previous version</summary>
![management-import-optionsets-1-changes](https://github.com/rdmorganiser/rdmo/assets/13996213/afec9fae-839b-496f-a817-5329f4167a45)
</details>



## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: --->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Code style update (formatting, renaming)
  - some type hints have been added to the new code
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [x] I have read the [contributor guide](https://github.com/rdmorganiser/rdmo/blob/master/CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
  - [x] also added more front end e2e tests
- [x] All new and existing tests passed.

## Still todo
Front-end
  - [x] add filter on top of import page
  - [x]  add info messages summary on top of import page
  - [x] button to show only the changed elements


<!--- This pull request template is adapted from:
<!--- "open-source-templates", https://github.com/TalAter/open-source-templates (MIT License). --->
<!--- "amazing-github-template", https://github.com/dec0dOS/amazing-github-template (MIT License). --->

MyPyDavid commented:
- [ ] fix the actions in the sidebar

jochenklar commented on package-lock.json:
is not up to date.

jochenklar commented on rdmo/conditions/imports.py:
```diff
@@ -1,42 +1,14 @@
-
-from rdmo.core.imports import check_permissions, set_common_fields, set_foreign_field, validate_instance
-
+from ..core.import_helpers import ElementImportHelper, ExtraFieldDefaultHelper
```
`rdmo.core.import_helpers`

jochenklar commented on rdmo/core/constants.py:
```diff
@@ -77,3 +77,23 @@
+    'comment',
+)
+
+RDMO_MODELS = {
```
This can be obtained from the model using e.g. `str(Attribute._meta)`

jochenklar replied:
Ah I see it was just moved.

jochenklar commented on rdmo/core/constants.py:
```diff
@@ -77,3 +77,23 @@
     "tib": {"base": 1024, "power": 4},
     "pib": {"base": 1024, "power": 5},
 }
+
```
I think this should be in `rdmo.management`.

jochenklar commented on rdmo/core/import_helpers.py:
```diff
@@ -0,0 +1,47 @@
+
+from django.db import models
+
+from rdmo.core.constants import ELEMENT_COMMON_FIELDS
```
Or can be declared here since it is not used else where.

MyPyDavid replied:
yes, moved it in here

jochenklar commented on rdmo/core/imports.py:
```diff
@@ -29,34 +38,147 @@ def handle_fetched_file(filedata):
+    for lang_code, lang_verbose_name, lang_field in get_languages():
+        name_code = f'{field_name}_{lang_code}'
+        name_field = f'{field_name}_{lang_field}'
+        # get_key = name_field if get_by_lang_field_key else name_code
```
comment

jochenklar commented on rdmo/management/assets/js/components/sidebar/ImportSidebar.js:
```diff
@@ -4,9 +4,21 @@ import isEmpty from 'lodash/isEmpty'
   const { elements, success } = imports
+
+   const {
+    // elementsImported,
```
comment

jochenklar commented on rdmo/management/constants.py:
This is not needed since `RDMO_MODEL_PATH_MAPPER` is only used once.

MyPyDavid replied:
it's still a constant and otherwise all the models need to be imported in the `viewsets.py` only for this mapping anyway. I could revert it to the previous version though..

jochenklar commented on rdmo/core/xml.py:
```diff
@@ -1,38 +1,125 @@
+
+
+@dataclass
+class XmlToElementsParser:
```
Why is this a data class and not a regular class?

MyPyDavid replied:
yes, you're right this was not ideal.
I've reverted and refactored this back into the functional approach.
Each function returns something like `obj, error` to keep the error handling consistent.

jochenklar commented on rdmo/management/imports.py:
```diff
@@ -1,66 +1,196 @@
+
+
+# mapping is redundant, since each ImportHelper has a .model_path attribute
+ELEMENT_IMPORT_HELPERS = {
```
please format dicts like this:

```
a = {
    ...
}
```

jochenklar commented on rdmo/management/imports.py:
```diff
@@ -1,66 +1,196 @@
+    original = copy.deepcopy(instance) if not _created else None
+
+    # prepare a log message
+    _msg = make_import_info_msg(model._meta.verbose_name, _created, uri=uri)
```
Why the `_`?

jochenklar commented on rdmo/management/imports.py:
```diff
@@ -1,66 +1,196 @@
+    if save:
+        logger.info(_msg)
+        instance.save()
+    # breakpoint()
```
clean plz

jochenklar commented on rdmo/management/imports.py:
I think this file needs some cleanup and some newlines to make it more readable.

MyPyDavid replied:
I have refactored the functions in the file a bit for readability and moved functions to `import_utils.py`

jochenklar commented:
I am sorry, there is more work to do on the css. Paddings and margins are off all over the place.

jochenklar commented:
![Screenshot_2024-05-16_15-08-24](https://github.com/rdmorganiser/rdmo/assets/1799675/fdd278d7-10a6-485e-bae4-f53334567dfc)

The padding needs to align with the rest of the panel.

jochenklar commented:
![Screenshot_2024-05-16_15-09-37](https://github.com/rdmorganiser/rdmo/assets/1799675/e097a919-2aa3-48db-9106-5687dc88dce6)

Same here around the warning. It think it 15px.

jochenklar commented:
![Screenshot_2024-05-16_15-11-46](https://github.com/rdmorganiser/rdmo/assets/1799675/563b7cf3-3252-45aa-8b8b-a2c9b37ea4be)

The different items need the same `margin-bottom`, also here the `margin-left` needs to be gone.

jochenklar commented:
![Uploading Screenshot_2024-05-16_15-15-19.png…]()

This needs more styling, the background should be white (except the red/green part), Current/plus/minus looks like they are clickable, but they are not. The whole block needs a margin-bottom of 10 or 20. The mono space font should be smaller.

jochenklar commented:
![Screenshot_2024-05-16_15-16-21](https://github.com/rdmorganiser/rdmo/assets/1799675/e56110c4-06f3-4993-b322-9b770b9f6ae2)

Here the indentation is off.

MyPyDavid commented:
> ![Screenshot_2024-05-16_15-09-37](https://private-user-images.githubusercontent.com/1799675/331267602-e097a919-2aa3-48db-9106-5687dc88dce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5Njc2NTgsIm5iZiI6MTcxNjk2NzM1OCwicGF0aCI6Ii8xNzk5Njc1LzMzMTI2NzYwMi1lMDk3YTkxOS0yYWEzLTQ4ZGItOTEwNi01Njg3ZGM4OGRjZTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjlUMDcyMjM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWM1ZWE3OTlhNzUyZjZlNWMxMjMxMmZiNTQ4M2Y3MDE2ZTVmODBjY2Y5ZmU2YWUwMTllNzdkNTkyZWVjYzFmMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.EVL6xRPt8hmgvX-Q5YBvxXjsgc9ERrkrqGVY11NmFKc)
> 
> Same here around the warning. It think it 15px.

thanks yes, I had `panel-body` `<div>` at the wrong place in the hierarchy of the component

MyPyDavid commented:
it looks somewhat better now I think, however there are still some todos...
![image](https://github.com/rdmorganiser/rdmo/assets/13996213/f59cc544-5d12-4af8-b0ee-153ac758c54c)

jochenklar commented:
Yes, much better, but there is now another panel around the entries, I would rather remove it. `panel-body` and `list-group` can be combined like this: https://getbootstrap.com/docs/3.4/components/#panels-list-group

jochenklar commented:
I would also remove the red and green background (it conflicts with the border radius of all other panels buttons etc.). The plus and minus look like buttons, you could remove the border and add the light red and blue there.

We should also remove the "row/col indentation" from the displayed fields. Maybe we put the field name on top, and display either the diff or just the fields value, a bit like this:

**help_de**
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.|

**help_en**
```
[+] Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
    invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
    et justo duo dolores et ea rebum.
[-] Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
    invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
    et justo duo dolores et ea rebum.
```

jochenklar commented:
Looking at it again, can you set the padding of the areas with the text and the +/- to something like: `2px 8px 2px 8px` and set the red green backround in the areas instead of outside?

MyPyDavid commented:
> I would also remove the red and green background (it conflicts with the border radius of all other panels buttons etc.). The plus and minus look like buttons, you could remove the border and add the light red and blue there.
> 
> We should also remove the "row/col indentation" from the displayed fields. Maybe we put the field name on top, and display either the diff or just the fields value, a bit like this:
> 
> **help_de** Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.|
> 
> **help_en**
> 
> ```
> [+] Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
>     invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
>     et justo duo dolores et ea rebum.
> [-] Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
>     invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
>     et justo duo dolores et ea rebum.
> ```

I think that showing only the diffs might be too confusing and maybe the user wants to sometimes copy/paste the values out of the page..

jochenklar commented:
ok, makes sense. The regular output is also markdown rendered, right?

MyPyDavid commented:
margin/paddings ok like this?
![image](https://github.com/rdmorganiser/rdmo/assets/13996213/742f9cfe-8755-42a6-89f8-f7e9b7b9db80)

MyPyDavid commented:
> ok, makes sense. The regular output is also markdown rendered, right?

need to look into it, haven't changed anything to the regular output I think

MyPyDavid commented:
- [ ] think that the `additional_input` field on `option` still has problem, it should show `[+] text` here
![image](https://github.com/rdmorganiser/rdmo/assets/13996213/2371f53b-cbf2-4625-b766-9eb25e49f8e6)

jochenklar commented:
Can you change this 
![image](https://github.com/rdmorganiser/rdmo/assets/1799675/a38f8cfa-aef9-4b4b-b421-e94d04f5ceb4)

```
Select all
Select changed
Deselect all
Deselect changed
```
Same for show. (I think it is deselect not unselect.)

jochenklar commented:
The `margin-top` from `.import-buttons` needs to go.

jochenklar commented:
One more: I think the checkbox should be `Show only created and changed` and should also include the elements to be created. The `.horizontal-container` needs to have `padding-bottom: 0;`

jochenklar commented:
Yes, but the indent looks somehow "off". In the interview we have the same, but there is a moving arrow so there it makes sense.

MyPyDavid commented:
I have one question about switching language/translations during import. It currently resets the page back to management (cancels the current import). 
How could I prevent this and just switch language during the import? Do I need to hook the `imports.elements` back into a reducer somewhere?

jochenklar commented:
I think its fine, there would be no straight forward way to keep the state when the page is reloaded.

MyPyDavid commented:
- [ ] the URIs for Attributes are not properly built at import (as shown in the `CodeLink` )
![image](https://github.com/rdmorganiser/rdmo/assets/13996213/0ddd92fa-b6b1-4e46-adec-4a9deaf2ac44)

MyPyDavid commented:
Front-end
- [ ] update some things in the `ImportSuccessElement`
- [ ] fix the placement in Warnings inside the panels

MyPyDavid commented:
I've implement the last todos (https://github.com/rdmorganiser/rdmo/pull/810#issuecomment-2155083683) and pasted in the current  [screenshot ](https://private-user-images.githubusercontent.com/13996213/340241837-e5fcb01c-bc8e-4c86-9b89-4318fd015c47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2MTQ4MTAsIm5iZiI6MTcxODYxNDUxMCwicGF0aCI6Ii8xMzk5NjIxMy8zNDAyNDE4MzctZTVmY2IwMWMtYmM4ZS00Yzg2LTliODktNDMxOGZkMDE1YzQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE3VDA4NTUxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjNzY5OTc3M2ZhNTc2MWVmM2RmYzBiNmEyZjkzNzI0NmIxMDRiZTliYTZjNmRhMDQxZDNhMWQxMTU5YTkxYzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6SfK2ZQqD83cbtcjIsCTaLgA5eJz_WZsA49zQy0WB74) 

Hopefully this is ready by now.
Tests could still be extended for the m2m cases...

MyPyDavid commented:
force pushed the interactive rebase on the current dev-2.2.0

jochenklar commented on rdmo/management/assets/js/components/import/ImportElement.js:
```diff
@@ -0,0 +1,82 @@
+          <ErrorLink onClick={updateShowField} />
+        }
+        {
+          (element.changed && element.updated) &&
```
This should not be this kind of link, it is confusing to users. We need something else to indicate that the element is created or updated. Before we used "created" in green. We could use `<span class="label label-success">create</span>`. It should also be on the left hand side after the URI.

jochenklar replied:
Looks great, tiny change: can we use `label-info` for `(element.changed && element.updated)`?

jochenklar reviewed (changes requested):
I changed small things with css. There are some issues regarding the warning list, I commented inline.

jochenklar commented on rdmo/management/management/commands/import.py:
```diff
@@ -15,16 +14,16 @@ def add_arguments(self, parser):
+            logger.info('Import failed with XML parsing errors.')
+            raise CommandError(str(e)) from e
+
+        # step 7: check if valid
```
I think this comment needs to be updated

MyPyDavid commented on rdmo/management/assets/js/components/import/common/Fields.js:
```diff
@@ -1,66 +1,58 @@
+        .sort()
+        .map(([key, value]) => {
+          if (!excludeKeys.includes(key)) {
+            return <FieldRow key={uniqueId()} element={element} keyName={key} value={serializeValue(value)} />
```
ok, however I would still show the empty field when it is in the `element.changedFields` though..

MyPyDavid commented on rdmo/management/assets/js/components/import/common/Warnings.js:
```diff
@@ -1,25 +1,47 @@
+
+  const prepareWarningsList = (warningsObj) =>
+    Object.entries(warningsObj).map(([uri, messages]) => (
+      <ul className="list-unstyled" key={uniqueId('warning-uri-list')}>
```
yes, tahnks, it was double `ul`

MyPyDavid commented on rdmo/management/assets/js/components/import/common/Warnings.js:
```diff
@@ -1,25 +1,47 @@
+    Object.entries(warningsObj).map(([uri, messages]) => (
+      <ul className="list-unstyled" key={uniqueId('warning-uri-list')}>
+        {shouldShowURI &&
+          <li key={uniqueId('warning-uri-')}>
```
yes, I've moved it to the `li` and adapted it to the element

MyPyDavid commented on rdmo/management/assets/js/components/import/common/Warnings.js:
```diff
@@ -1,25 +1,47 @@
+  }
 
-export default Warnings
+  export default Warnings
```
I don't understand what you mean here

jochenklar replied:
Ah, sorry, I meant to use the same markup and style for the list at the top of the page and the warnings for each element including the yellow panel. Then we don't need the "Warning" headline which looks a bit lost.

If we use https://getbootstrap.com/docs/3.3/components/#panels-list-group the whole list would be more readable.

(I just put the comment at the end of the file, which was not so smart, I guess.)

MyPyDavid replied:
ok, I've used the `list-group` for the Warnings and Errors. 
Refactored the components and props to be reusable in the warnings of an element and in the aggregate at the top of the page.
Think it looks better now, nearly finished right?

MyPyDavid commented on rdmo/management/assets/js/components/import/common/Warnings.js:
```diff
@@ -1,25 +1,47 @@
+  const listWarningMessages = prepareWarningsList(element.warnings)
+
+  return (
+    <div className="row text-warning mt-10">
```
without the `row` and `col-` the padding/margin of the Warning text doesn't align with the fields any more, or would need some other css class..

jochenklar replied:
I just removed them and it looks fine, but maybe I am missing something. Or you removed only `col-` but not `row`.

MyPyDavid replied:
ah thanks, you're right. The empty `<div>` looks so strange..
I've removed them from `Warnings.js` and `Errors.js`

jochenklar commented:
Alright! I think we are done. I adjusted the margings and changed the ShowLink icon to chevron-up/down. I hope thats ok with you. 
![image](https://github.com/rdmorganiser/rdmo/assets/1799675/42af6a47-a9c1-41ec-9b67-4c6c4dbafa91)

MyPyDavid commented:
thanks for the review and changes! 
I think I still found one open to-do:
* [ ] fix changing the path/key when importing an Attribute.

MyPyDavid commented on rdmo/management/assets/js/reducers/importsReducer.js:
```diff
@@ -48,6 +48,9 @@ export default function importsReducer(state = initialState, action) {
         const elements = [...state.elements]
         elements[index] = {...elements[index], ...action.values}
+        if (elements[index].model === 'domain.attribute') {
+          elements[index].path = buildPath(elements[index].key, elements[index].parent ? elements[index].parent.uri : null, elements)
```
maybe `buildPath` should be called `buildPathForAttribute`  when it's only used for the attributes?

MyPyDavid commented on rdmo/management/assets/js/utils/elements.js:
```diff
@@ -174,4 +174,24 @@ const buildUri = (element) => {
+
+  let path = key
+  if (parentUri) {
+    const parentElement = elements.find(element => element.uri === parentUri)
```
I've added this find `path` from another parent element but I'm not sure if it's needed here to build the path only for attributes.

Would only this `parentPath = parentUri.split('/domain/')[1]` not be sufficient to build the path for a single attribute element?

jochenklar replied:
Yes, looking at this function, and if its only needed for Attributes, then I would simplify the function like you described.

jochenklar reviewed (changes requested):
I found something else which is rather small.

![image](https://github.com/rdmorganiser/rdmo/assets/1799675/5a809fb4-eb45-472b-b050-231f257a0790)

The `.` needs to go, and on this view it should not be `created` and `updated` since the import was not performed yet. Maybe `create` and `update`?

jochenklar commented:
Same for `changed` and also in the filter.

MyPyDavid commented:
> I found something else which is rather small.
> 
> ![image](https://private-user-images.githubusercontent.com/1799675/345046288-5a809fb4-eb45-472b-b050-231f257a0790.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5ODk2MjcsIm5iZiI6MTcxOTk4OTMyNywicGF0aCI6Ii8xNzk5Njc1LzM0NTA0NjI4OC01YTgwOWZiNC1lYjQ1LTQ3MmItYjA1MC0yMzFmMjU3YTA3OTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDNUMDY0ODQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjc3NjYyNzAxMzU3YWJiODk1OGQ0NGVlYTJmNGMyMzYwYTZmMWM1OTMzYWZmYzY2NGJlY2ZlZTQwOWI4NjYyMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1IgGm7YFfDeRzuhr5wJjSw_T4FJ9Z8KlDbxTSk38Fa0)
> 
> The `.` needs to go, and on this view it should not be `created` and `updated` since the import was not performed yet. Maybe `create` and `update`?

I've updated this in the labels and in the filter label. Only the "create" looks a bit strange. As alternative, maybe "new" or "creation" but am also not sure about those..

MyPyDavid commented:
rebased to dev-2.2.0 and force pushed

jochenklar commented:
I think new is great since we use new/neu also in the interface.

MyPyDavid commented:
ok, changed the text to "new"
and added (new from dev-220) the `short_title` field to the import helpers

jochenklar commented:
Sorry I found more bugs:

* if you upload and import a faulty file the message is missleading:

![image](https://github.com/rdmorganiser/rdmo/assets/1799675/a5cfa16f-70d2-4d49-b8b2-f85993a8539b)

with this xml
```
<?xml version="1.0" encoding="UTF-8"?>
<rdmo xmlns:dc="http://purl.org/dc/elements/1.1/" created="2024-07-01T10:53:07.185903+02:00" version="2.2.0.dev1">
	<question dc:uri="http://example.com/terms/questions/catalog/blocks/set/block/b">
		<uri_prefix>http://example.com/terms</uri_prefix>
	</question>
</rdmo>
```

(GitHub does not allow XML uploads)

* For the same file: I got this

![image](https://github.com/rdmorganiser/rdmo/assets/1799675/ff3b411b-744a-46ec-b183-207b1e2aba12)

after I click **Question** or click show. I think this is also because of the broken xml file, so it is not super important.

* I would remove the onClick on the created/updated badge or add `cursor: pointer;` using css, upt to you.

MyPyDavid commented:
thanks for noticing and the xml.

>     * if you upload and import a faulty file the message is missleading:

 This came from the back-end. I've changed it now so that "new" and "create" will not show at import when there are errors (and the db instance won't be saved anyways).

>     * I would remove the onClick on the created/updated badge or add cursor: pointer; using css, upt to you.

I've removed the onClick from the badge.

One other issue with this faulty xml was that the `buildUri` doesn't function properly anymore (it shows `../undefined` in the URI).
I've changed it now so that the front-end that it will keep the `element.uri` when `buildUri` can not build the URI and returns `null`.

Another mini-feature addition I thought about was, to make these URIs at import also clickable (so that they open the edit page of each element). I would just need to the `element.id` from the backend and build the link. Do you think I should add that here?

jochenklar commented:
No I would not add that. It could lead to problems if the users then go back in the browser. They should just click on "Back".
